### DISCUSSION TITLE
How to use Start In Relative

### DISCUSSION THREAD

**Discussion by Murtang** [ekutner/home-connect-hass#114] [Community Support] [answered]
Probably an easy answer to this question but I can't seem to figure it out.
I have a Bosch dishwasher and connected it through Home Connect Alt in HA. I can start it directly from the app and also I can set the start in relative. I see that the dishwasher starts counting down, but when it has finished the count down, it does not start automatically.

Is there a setting that should be changed for this, or shall I make an automation to detect when the countdown is down and then call the  start function?

**Comment by MBakerSE** [comment]
Start in relative is just a counter. After you have set this you actually have to start the dishwasher. I run a button.press service for the start_pause button on my Siemens after I have set Start in relative. 

Hope it helps!

**Reply by ChristophCaina** [reply]
one question: shouldn't it be vice versa?
First triger "start_pause" -> and then the start_in_relative counter should come in and start the machine when the counter is down to 0 ?

I don't have such a device, but just having a counter option on the device - and then require another start_pause action doesn't seem to be the way it is supposed to be

**Reply by MBakerSE** [reply]
That is exactly what I meant. You first set the timer, then you press play before timer gets to zero (i.e. right away), then when counter gets to zero the program will start. If you do not press pay, nothing happens when timer gets to zero. 

I didn't mean you should wait for the timer to finish 😀.

**Reply by Murtang** [reply]
Thank you both for your responses. I'll try it as soon as I find time to play around with it a little more. Also working on another part where it shall find the next 2 or 4 hour interval where the electricity is cheapest which it will be combined with.

**Reply by MBakerSE** [reply]
This is exactly what I do. I use Nordpool integration to find the cheapest price in Sweden and I have logged the power consumption of my dishwasher. Then I run a python script when I click the "remote control" button on the dishwasher to calculate optimal time to start.

**Reply by Murtang** [reply]
Hi @MBakerSE 

Finally got some time to get further with this. And also finally understood/found the button press service. Still a lot in HA I need to learn.

I have a test script where first I run Entity - Set value for Bosch Opvaskemaskine - start in relative and thereafter run the button press Bosch opvaskemaskine Start. See picture below. But after the countdown, nothing happens. Am I still missing something compares to how you explain you do it?
![Screenshot_2022-11-29-22-18-15-40_c3a231c25ed346e59462e84656a70e50](https://user-images.githubusercontent.com/1453380/204650221-3d851cce-a3e8-479e-8137-11e0105afc1c.jpg)

**Reply by MBakerSE** [reply]
Did you select a program first, so it knows what to run? For Siemens one is always selected I think, but not sure if it's the same on Bosch. There should be a select for the programs.

**Reply by Murtang** [reply]
Tried that also, no difference. Really starting to get frustated, either I am a complete idiot, or my Bosch machine does not behave similar to the siemens.

Can I ask you to test where in your sequence, that the Operating State is changed to Delayed Start. I am starting to think that my problem is that it never gets set to Delayed Start but stays at Ready

**Reply by MBakerSE** [reply]
Yes, mine goes to delayed start - when I "press" the start button I think. Have you tried to run the start button press without setting start in relative? Then the mashine should start immediately. The start in relative will count down no matter if you press start or not, so it might be a problem with the button start never getting triggered.

**Reply by MBakerSE** [reply]
Stupid question perhaps, but have you activated remote start by pressing the physical remote connect button on the washer? Otherwise you cannot start it remotely. 
I have it set so that when remote control start allowed goes to on, my automation (and pythonscript) is run.

**Reply by Murtang** [reply]
Yes, I can start it directly with the start button fro HA. Stop also works, just haven't figured out the delayed start correct setup yet.
Would you mind sharing your python script? Then I could try that? Maybe there is a difference between Siemens and Bosch behavior

**Reply by MBakerSE** [reply]
The Python script only sets the Start in relative and which program to run, the I press start just as you do with an automation. So if you have managed to actually set a start in relative number and a program,  then this is not the problem.

One funny thing I noticed though in your picture.
You have a Norwegian UI. In mine, in the visual editor, it says that the service is "knapp: press" (Swedish). Should be the same for you I think?
Have you checked that the YAML code of you service call corresponds to
`service: button.press
data: {}
target:
  entity_id: button.[...long number...]_start_pause
`
(Replaced the id number of my entity)

**Comment by ekutner** [comment]
The appliance should start automatically when setting it to start with a delay. This is not being handled by the integration directly, it just sends the information to the appliance with the start program command and the appliance is supposed to do it by itself, the same way that would work if you configured a delayed start directly at the appliance.

**Reply by MBakerSE** [reply]
Actually on the applience you first set the delay and then press play to "activate it". So it is the same.
There is one annoying thing about the "Start in relative" though and it is that it becomes unavailable if you restart home assistant. The only way to get it available again is to either run a program manually after having set a delay on the device, OR switch to another program. Switching program back and forth is a work around, but it seems unnecessary.
I might write a bug on it later, but it's no big deal...

@ekutner: Amazing work with this integration! It works beautifully.

**Reply by Speedy1496** [reply]
Hello,
I have the same issues :
- The "Start in relative" didn't appear after rebooting and I had to change the program to make it available
- If I set a value in "Start in relative" the "Stop" become available and the "Start" become unavailable, the count down go to 0 and nothing happen.
- If "Start in relative" is not 0 I can't press "Start"

I actually have only the solution to wait the "Start in relative"  to go to 0 then press "Start" with an automation, 

If I use the google app "Home Connect" on my phone, I can start the Appliance delayed remotely.

I'm on the Bosch French version SMS68TW16E
Thank you 
Sébastien

**Reply by Speedy1496** [reply]
I wonder if we have to use the "Home Connect Alt: Start program" with Options...

**Reply by Murtang** [reply]
Maybe there just is a difference between the api to the Bosch and Siemens dishwashers...

I have come up with the same solution as you for now. I have an automation that detects when the door is closed and remote has been activated and then runs a script to set the Start in relative timer.

Then I have another automation which condition is that Start in relative changes to 0, then it calls the Start function

**Comment by ekutner** [comment]
See my previous comment on the expected behavior, the integration isn't controlling the process. it just sends the command to HC . I will try to reproduce the issue you are seeing on my own appliance but can't say when I'll have the time to do it.

**Comment by ekutner** [comment]
I've tested this again and it is still working ok for me. I set the "Delayed operation" selection then start the program and my washing machine starts working when it gets to the correct time. Note that the timer is setting when the program will finish, not when it will start, so if you set it to 3:00h and the program is 1:20h then it will wait 1:40h and then start the program. This all happens automatically, the integration doesn't manage it beyond sending the initial delayed operation command to HC.

**Reply by kenni** [reply]
I've just setup the integration and I'm unfortunately experiencing the same issue on a Bosch SMV8YCX01E. I'm currently trying to figure out any differences between doing a delayed start from the integration and from the appliance. One thing I noticed so far going through the logs:

On february 7 around 10:00, I scheduled a delayed start, setting "Start In Relative" so its counter would hit 0 an hour later, on February 7 at 11:00:15. The Start-button becomes unavailable, so it I cannot press "Start", but according to the Home Connect app and the dishwasher, it claims it will run at the specified time. When the time is up and it should start, the state of the Start-button changes from Unavailable to "February 6, at 21:45" which corresponds to the night before, when I previously started a cleaning:
<img width="236" alt="image" src="https://user-images.githubusercontent.com/486971/217227442-56cb2b1e-b8de-49d0-ae0d-000fc417f30d.png">

Immediately thereafter, the Operating State changes to "Ready" and nothing happens:
<img width="234" alt="image" src="https://user-images.githubusercontent.com/486971/217229198-8391769d-bf3d-40c1-b252-27d6a3df3b5d.png">

When I then go to the machine and manually press on the Start-button, the state changes to the current timestamp and the machine starts the program:
<img width="236" alt="image" src="https://user-images.githubusercontent.com/486971/217229614-a290ff96-3d2a-4448-9e9a-9644e2059d7f.png">

<img width="236" alt="image" src="https://user-images.githubusercontent.com/486971/217248385-018c7a24-d11d-4ecf-b31d-05e9ff136199.png">

It might just be a result of the (firmware/API?) bug, but it stood out to me during first glance at the logs.

**Reply by ekutner** [reply]
How are you setting it, are you using an automation? If you are then don't, first make sure it is working correctly when done manually in the UI.
Also, note that the time should be the time until the program **finishes**, not the time it starts.

**Reply by kenni** [reply]
It doesn't matter if I do it directly in the HA webinterface or if I do it through an automation in Node Red. I'll try to go through the steps below. I also cannot replicate that the time should be when the program finishes, as I'll try to illustrate below:

First I power on the dishwasher and select the Eco50-program:

<img width="497" alt="01" src="https://user-images.githubusercontent.com/486971/217368018-85c98b22-bfae-4712-ab62-113dedbbb48e.png">

The clock is 22:10 now and the Eco50 program takes about 4 hours and 10 minutes to run. As a test I enter 18000 seconds (5 hours) into the "Start in relative" entity ("number.011090517384007772_bsh_common_option_startinrelative") and press on enter:

<img width="488" alt="02" src="https://user-images.githubusercontent.com/486971/217369854-ab7ee69d-b76d-4c6a-aba1-6498b3089094.png">

The Start-button gets greyed out and the Operating State in HA changes to "Delayed Start". When I now go and check the dishwasher (it projects the time of program completion onto the floor), it has set the complete time to 9 hours and 10 minutes from from now (4 hours 10 minues + 18000 seconds), meaning that the number I entered must have been the start time of the program:
![03](https://user-images.githubusercontent.com/486971/217370842-2fab817a-501d-4574-ac3e-ab192705ae2d.jpeg)

**Reply by ekutner** [reply]
You're right. I just now realized that my appliance has Finish In Relative and you are talking about Start In Relative. So ignore everything I said, I was just to set on the way my appliance does it that I missed how it said "start in relative" in the thread title...

However, if that is the case then I have no idea what to expect. If someone could send a debug log with log_mode: 7 and a press on the Home Connect Debug button before before and after starting the program with the delayed start, that could help.

**Reply by kenni** [reply]
I'll be happy to provide any debug logs needed. When I read this project's README or the HA logger documentation, I'm unable to find the "log_mode" argument. Where do you specify that? I've copy-pasted the logger-section from the README.

**Reply by kenni** [reply]
Ok, assuming that the "log_mode: 7" should be placed below the "home_connect_alt"-section and not the "logger"-section. I've attached the log below.

I perform the following actions:
1. Press on the debug button
2. Turn on the dishwasher
3. Select the Auto-program (I believe that it is called "Dishcare.Dishwasher.Program.Auto2")
4. Set the "Start in Relative" to 60 seconds
5. Wait for the 60 seconds to pass and the dishwasher to change to "Ready" state (where it should change to "Run" state)
6. Press on the debug button again

[home-assistant_2023-02-07T22-04-05.479Z.log](https://github.com/ekutner/home-connect-hass/files/10679800/home-assistant_2023-02-07T22-04-05.479Z.log)

**Comment by ekutner** [comment]
I'm not sure what is going on there and can't test it on my own appliance. As far as I can tell from the log you never actually pressed the Start Program button.
Let's try it another way,  using the Start Program service. Set the program you want (Dishcare.Dishwasher.Program.Auto2 for example) and in the program options specify BSH.Common.Option.StartInRelative: 180 (I intentionally want a 3 min delay)
Then send me the log.

**Reply by kenni** [reply]
That works!!!...and yes, I can confirm that I do not press on the Start Program button after setting Start In Relative. As previously described, the button becomes unavailable/greyed out milliseconds after setting "Start In Relative", so it is impossible to press the button. Another user, @Speedy1496, also describes the same behaviour in a previous comment in this thread (https://github.com/ekutner/home-connect-hass/discussions/114#discussioncomment-4768461).

Logs attached:
[home-assistant_2023-02-11T13-26-48.617Z.log](https://github.com/ekutner/home-connect-hass/files/10713769/home-assistant_2023-02-11T13-26-48.617Z.log)

**Comment by ekutner** [comment]
Ok, that's progress.
Try installing version 0.6.0-b4 and see if it works using the UI. If it doesn't then upload the log again. Note tat in the UI select box you can only select in 30 min increments.

**Comment by kenni** [comment, marked as answer]
Definitely progress :)

With 0.6.0-b4 using the GUI, I can:

- Power on the dishwasher
- Select the program

But I cannot set any "Start In Relative" time, as it is unavailable. I've tried restarting HA a couple of times, and powering the dishwasher on and off, but no change.

<img width="568" alt="Screenshot 2023-02-11 at 15 43 09" src="https://user-images.githubusercontent.com/486971/218264438-9b179bf8-9a0b-42ac-8111-4d08997f99b0.png">

I noticed that the related sensor (sensor.011090517384007772_bsh_common_option_startinrelative) has changed it format from amount of seconds ("0") to:
"0:00"

Logs attached below:
[home-assistant_2023-02-11T14-45-09.440Z.log](https://github.com/ekutner/home-connect-hass/files/10713907/home-assistant_2023-02-11T14-45-09.440Z.log)

**Reply by ekutner** [reply]
The number entity is gone, that is ok, you should instead have a new select box entity with the same name

**Reply by kenni** [reply]
Ohhh...I'll be back in 31 minutes :) Thanks.

If this works, will the start program option with "BSH.Common.Option.StartInRelative: 180" (in seconds) still be supported? I see it as a huge advantage to be able to specify the start time in seconds from automations.

**Reply by kenni** [reply]
Yes, it works through the GUI now as well :) ...and I can see that I can still call the start program service with the amount of seconds, so I think everything is good now. Thanks a lot for your help and your work on the integration.

Log is attached if any verification is needed:
[home-assistant_2023-02-11T15-37-15.849Z.log](https://github.com/ekutner/home-connect-hass/files/10713996/home-assistant_2023-02-11T15-37-15.849Z.log)

**Reply by ekutner** [reply]
Great. I hope this wraps up the issue for everyone.

**Reply by knights1** [reply]
@ekutner : bit offtopic, but I used the FAVORITE button which is on the dishwasher to 'start' an automation which has a time set for (00:00) for starting the appliance. So I was not using start relative.

But since the last 0.6 version I no longer see the entity "BSH.Common.Program.Favorite.001", is there somewhere in which I can get this back? Its great having a physical button to play with and use HA for automation.



@kenni : perhaps the FAVORITE button on your dishwasher can be of use to. Then you only press 1 button and everything is taken care of. Can imagine that if you have Dynamic electricity contract that you could optimize this when costs are lowest in the night/day.

**Reply by ekutner** [reply]
The changes in 0.6.0 shouldn't have an impact on which programs are seen by the integration.
Please enable debug logging, as described in the README, and add log_mode: 7 to the home_connect_alt section in the config file. Then reproduce the issue and upload the file in a new github issue report (not here).
